proposals/1113_ldap-ad-web-ui/README.md

```diff
@@ -0,0 +1,136 @@
+Therefore, we need a mechanism for untrusted clients to acquire Pinniped's downstream tokens while
+leaving the IDP credential handling to the Pinniped supervisor.
+
+## Proposal
+Pinniped must provide a simple login screen in order to support UIs that wish
+to authenticate with the Pinniped Supervisor to gain access to a cluster without
+requiring each app to handle IDP credentials.
+
+### Goals and Non-goals
+
+#### Goals
+* Prevent OAuth clients, other than the Pinniped CLI, from providing credentials via the authorization request
+* Provide a minimal feature set (ie user id, password & submit button only)
+* Provide generalized error messaging for failed logins that do not expose sensitive information (i.e. we should say "invalid username or password"
+  but do not expose whether it's the username or password that's incorrect)
+* Provide information easily allowing a user to identify the screen as belonging to Pinniped and which upstream IdP is being represented (e.g. IdP name)
+* Address basic security concerns for web firms (HTTPS, passwords use a password field, CSRF protection, redirect protection)
+* Prevent LDAP injection attacks
+* Rely on the upstream IdP to address advanced security concerns (brute force protection, username enumeration, etc)
+* Screens are accessible and friendly to screen readers
+* Screens are friendly to password managers
+
+#### Non-goals
+* A rich client (ie the use of javascript)
+* Advanced UI features (e.g. remember me, reveal password). These features are better left to identity providers to implement.
+* Branding & customization beyond the information listed in the goals used to identify the login screen belongs to Pinniped.
+* Supporting SSO integrations
+* Internationalization or localization. The CLI doesn't currently support this either.
+
+### Specification / How it Solves the Use Cases
+
+#### API Changes
+
+The supervisor must accept requests from other clients, as detailed
+in the (todo) proposal for dynamic client registration.
+When a client other than pinniped-cli makes an authorization endpoint request with `response_type=code` and their
+IDP is an LDAP or Active Directory IDP, the user will be redirected to the new login page.
+When a client other than the Pinniped CLI makes an authorization endpoint request with
+custom Username/Password headers, they should be rejected.
+
+The pinniped cli should default to using the cli-based password flow, but when a tty is unavailable or when
+a CLI flag (`--ldap-browser-login`? `--force-browser-login`?) is specified, it will open a browser to log in
+instead of prompting for username and password. Some users (for example, IDE plugins for kubernetes)
+may wish to authenticate using the pinniped cli but without access to a terminal.
+
+#### Upgrades
+
+This change is backwards compatible. Users would see no changes unless they decided to register
+a new client or change the pinniped cli flags.
```

Comment:
I would expect the discovery metadata to change to state that the LDAP/AD IDPs also support `browser_authcode` now.  Newer CLIs should fallback to this option when there is no TTY/env vars set.  I think more of the flow logic needs to be moved into the `login` command and out of the `get kubeconfig` command (to allow this decision to be made at runtime).

Reply:
> I would expect the discovery metadata to change to state that the LDAP/AD IDPs also support browser_authcode now.

Yes. The `pinniped get kubeconfig` command already supports this by reading the results of the IDP discovery endpoint and allowing whatever the results of that endpoint dictate. So kubeconfigs generated by both old and new CLIs would be the same.

An old CLI will output a friendly validation error during login if the kubeconfig chooses `browser_authcode` for an LDAP/AD IDP. This would be solved by upgrading the CLI to a version which skips that validation. That seems like reasonable upgrade behavior. It could be documented in the release notes.

> I think more of the flow logic needs to be moved into the login command and out of the get kubeconfig command (to allow this decision to be made at runtime).

The runtime decision to prefer the `browser_authcode` flow over the `cli_password` flow could be done with a new environment variable which forces the use of `browser_authcode` regardless of the setting of `--upstream-identity-provider-flow` from the kubeconfig. That would be sufficient to handle use cases like IDE integrations which don't have access to a terminal to prompt for usernames/passwords and need to override the flow selection. That could be part of this proposal, or could be a separate small feature after this proposal is implemented.

In the future, we could offer more capabilities to optionally move the selection of IDP and flow into the end user's runtime choice during login, however that's probably beyond the scope of this proposal. For the purposes of this proposal, the CLI's existing implementation should be good enough.

Reply:
See also https://github.com/vmware-tanzu/pinniped/pull/1132#pullrequestreview-948919721 regarding the behavior of `pinniped get kubeconfig`.

Reply:
I think this conversation was resolved in offline discussions and the doc was updated accordingly.